### DISCUSSION TITLE
Add Promise.withResolvers() Polyfill to Next.js

### DISCUSSION THREAD

**Discussion by isBatak** [vercel/next.js#71027] [Ideas]
### Goals

1. Add support for `Promise.withResolvers()` in Next.js for both the browser and Node.js environments.  
2. Ensure compatibility with Next.js's supported browser versions, particularly Safari versions between 12 and 17.4
   - Chrome 64+  
   - Edge 79+  
   - Firefox 67+  
   - Opera 51+  
   - Safari 12+  
3. Enable `Promise.withResolvers()` polyfill for Node.js, especially for versions before Node 22.
4. Ensure that Promise.withResolvers() is supported in TypeScript 5.4, which already has built-in support for this method.

### Non-Goals

1. Support for older browsers not officially supported by Next.js.

### Background

`Promise.withResolvers()` is a useful method that was added to the JavaScript specification and is expected to be baseline-supported by 2024, as documented in [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers#description). However, as of now, this feature is only supported in Node.js version 22 and above. Since Next.js supports a range of browsers starting from Chrome 64, Edge 79, Firefox 67, Opera 51, and Safari 12, a polyfill would help ensure compatibility across both browser and server environments. Adding this polyfill can smooth the transition for developers until all supported environments natively support `Promise.withResolvers()`.

### Proposal

Add the polyfill for `Promise.withResolvers()` in both the Next.js polyfill modules:  
- [`next-polyfill-module`](https://github.com/vercel/next.js/blob/canary/packages/next-polyfill-module/src/index.js)  
- [`next-polyfill-nomodule`](https://github.com/vercel/next.js/blob/canary/packages/next-polyfill-nomodule/src/index.js)
- Replace [detached-promise](https://github.com/vercel/next.js/blob/b9617dbb4ec35e8350bec5892e950210d738d449/packages/next/src/lib/detached-promise.ts#L4) implementation with Node polyfill

**Comment by timneutkens** [comment]
Replied in thread: https://x.com/timneutkens/status/2024049570351546764

Next.js doesn't include every polyfill possible as it would increase bundle size of all applications. 

The browser support level means that is the minimum amount of APIs available. The API you're talking about was added after the browser support level, so it's not included. Does that make sense?

For the case you're explaining adding polyfills is up to you. 

Can use [instrumentation-client](https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client) for that 👍
